Implement input validation for CSSRGB constructor and setters

cdumez opened the pull request:
#### c4f1f9c6e5d1a6085ded4dfbc458a965024ed8a4
<pre>
Implement input validation for CSSRGB constructor and setters
<a href="https://bugs.webkit.org/show_bug.cgi?id=246972">https://bugs.webkit.org/show_bug.cgi?id=246972</a>

Reviewed by Antti Koivisto.

Implement input validation for CSSRGB constructor and setters:
- <a href="https://drafts.css-houdini.org/css-typed-om-1/#dom-cssrgb-cssrgb-r-g-b-optional-alpha">https://drafts.css-houdini.org/css-typed-om-1/#dom-cssrgb-cssrgb-r-g-b-optional-alpha</a>
- <a href="https://drafts.css-houdini.org/css-typed-om-1/#rectify-a-csscolorrgbcomp">https://drafts.css-houdini.org/css-typed-om-1/#rectify-a-csscolorrgbcomp</a>
- <a href="https://drafts.css-houdini.org/css-typed-om-1/#rectify-a-csscolorpercent">https://drafts.css-houdini.org/css-typed-om-1/#rectify-a-csscolorpercent</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssRGB-expected.txt:

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssRGB.html:
Made some fixes to the test, which I will export to WPT:
- Expect a Syntax error instead of a TypeError, as per the specification.
- The loop for iterating over invalid test cases for the setters was wrong. As a result, the setters kept
  getting called with the string &quot;undefined&quot; instead of the actual test value.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/typedom/color/CSSColorValue.cpp: Added.
(WebCore::CSSColorValue::colorSpace):
(WebCore::CSSColorValue::to):
(WebCore::CSSColorValue::parse):
(WebCore::CSSColorValue::rectifyCSSColorPercent):
(WebCore::CSSColorValue::toCSSColorPercent):
* Source/WebCore/css/typedom/color/CSSColorValue.h:
(WebCore::CSSColorValue::colorSpace): Deleted.
(WebCore::CSSColorValue::to): Deleted.
(WebCore::CSSColorValue::parse): Deleted.
* Source/WebCore/css/typedom/color/CSSRGB.cpp:
(WebCore::toCSSColorRGBComp):
(WebCore::CSSRGB::create):
(WebCore::CSSRGB::CSSRGB):
(WebCore::CSSRGB::r const):
(WebCore::CSSRGB::setR):
(WebCore::CSSRGB::g const):
(WebCore::CSSRGB::setG):
(WebCore::CSSRGB::b const):
(WebCore::CSSRGB::setB):
(WebCore::CSSRGB::alpha const):
(WebCore::CSSRGB::setAlpha):
(WebCore::CSSRGB::rectifyCSSColorRGBComp):
* Source/WebCore/css/typedom/color/CSSRGB.h:
(WebCore::CSSRGB::create): Deleted.
(WebCore::CSSRGB::r const): Deleted.
(WebCore::CSSRGB::setR): Deleted.
(WebCore::CSSRGB::g const): Deleted.
(WebCore::CSSRGB::setG): Deleted.
(WebCore::CSSRGB::b const): Deleted.
(WebCore::CSSRGB::setB): Deleted.
(WebCore::CSSRGB::alpha const): Deleted.
(WebCore::CSSRGB::setAlpha): Deleted.

Canonical link: <a href="https://commits.webkit.org/256024@main">https://commits.webkit.org/256024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a767698b3dd38e1752d61bcb6291fe100155b66

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94090 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24630 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103715 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164055 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3297 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31502 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86398 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99755 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99749 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2370 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80509 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29378 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84296 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/83978 "Found 37 new API test failures: TestWebKitAPI.PDFHUD.MainResourcePDF, TestWebKitAPI.WKWebViewCloseAllMediaPresentations.ElementFullscreen, TestWebKitAPI.WebKit.MediaBufferingPolicy, TestWebKitAPI.WebKit.AutoplayOnVisibilityChange, TestWebKitAPI.WebKit.QuotaDelegate, TestWebKitAPI.CloseWebViewAfterEnterFullscreen.VideoFullscreen, TestWebKitAPI.VideoControlsManager.VideoControlsManagerMultipleVideosSwitchControlledVideoWhenScrolling, TestWebKitAPI.WKNavigation.AutomaticVisibleViewReloadAfterWebProcessCrash, TestWebKitAPI.CloseWebViewAfterEnterFullscreen.ElementFullscreen, TestWebKitAPI.WebKit.GeolocationPermission ... (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72330 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37888 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17801 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35769 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19068 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39643 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41641 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1969 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41584 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38301 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/typedom/color/CSSColorValue.cpp:
```diff
@@ -0,0 +1,81 @@
+#include "CSSUnits.h"
+#include "ExceptionOr.h"
+
+namespace WebCore {
+
+RefPtr<CSSKeywordValue> CSSColorValue::colorSpace()
```
It would be nice to move all the Typed OM types to a namespace of their own. The names are confusingly similar to the internal CSS types.

cdumez replied:
Good idea. I'll follow-up.

cdumez committed:
Implement input validation for CSSRGB constructor and setters
https://bugs.webkit.org/show_bug.cgi?id=246972

Reviewed by Antti Koivisto.

Implement input validation for CSSRGB constructor and setters:
- https://drafts.css-houdini.org/css-typed-om-1/#dom-cssrgb-cssrgb-r-g-b-optional-alpha
- https://drafts.css-houdini.org/css-typed-om-1/#rectify-a-csscolorrgbcomp
- https://drafts.css-houdini.org/css-typed-om-1/#rectify-a-csscolorpercent

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssRGB-expected.txt:

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssRGB.html:
Made some fixes to the test, which I will export to WPT:
- Expect a Syntax error instead of a TypeError, as per the specification.
- The loop for iterating over invalid test cases for the setters was wrong. As a result, the setters kept
  getting called with the string "undefined" instead of the actual test value.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/typedom/color/CSSColorValue.cpp: Added.
(WebCore::CSSColorValue::colorSpace):
(WebCore::CSSColorValue::to):
(WebCore::CSSColorValue::parse):
(WebCore::CSSColorValue::rectifyCSSColorPercent):
(WebCore::CSSColorValue::toCSSColorPercent):
* Source/WebCore/css/typedom/color/CSSColorValue.h:
(WebCore::CSSColorValue::colorSpace): Deleted.
(WebCore::CSSColorValue::to): Deleted.
(WebCore::CSSColorValue::parse): Deleted.
* Source/WebCore/css/typedom/color/CSSRGB.cpp:
(WebCore::toCSSColorRGBComp):
(WebCore::CSSRGB::create):
(WebCore::CSSRGB::CSSRGB):
(WebCore::CSSRGB::r const):
(WebCore::CSSRGB::setR):
(WebCore::CSSRGB::g const):
(WebCore::CSSRGB::setG):
(WebCore::CSSRGB::b const):
(WebCore::CSSRGB::setB):
(WebCore::CSSRGB::alpha const):
(WebCore::CSSRGB::setAlpha):
(WebCore::CSSRGB::rectifyCSSColorRGBComp):
* Source/WebCore/css/typedom/color/CSSRGB.h:
(WebCore::CSSRGB::create): Deleted.
(WebCore::CSSRGB::r const): Deleted.
(WebCore::CSSRGB::setR): Deleted.
(WebCore::CSSRGB::g const): Deleted.
(WebCore::CSSRGB::setG): Deleted.
(WebCore::CSSRGB::b const): Deleted.
(WebCore::CSSRGB::setB): Deleted.
(WebCore::CSSRGB::alpha const): Deleted.
(WebCore::CSSRGB::setAlpha): Deleted.

Canonical link: https://commits.webkit.org/256024@main
